test/extended/dr/common.go

```diff
@@ -543,3 +529,166 @@ func createPrivatePublicSSHKeySecret(oc *exutil.CLI, name string, namespace stri
+	podSpec.Containers = []applycorev1.ContainerApplyConfiguration{
+		*applycorev1.Container().
+			WithName("cluster-backup").
+			WithSecurityContext(applycorev1.SecurityContext().WithPrivileged(true)).
+			WithImage(image.ShellImage()).
+			WithVolumeMounts(
+				applycorev1.VolumeMount().WithName("keys").WithMountPath(sshPath),
+			).
+			WithCommand("/bin/bash", "-c", backupScript),
+	}
+	podSpec.NodeSelector = map[string]string{"kubernetes.io/hostname": backupNode.Labels["kubernetes.io/hostname"]}
+	podSpec.Tolerations = []applycorev1.TolerationApplyConfiguration{
+		*applycorev1.Toleration().WithKey("node-role.kubernetes.io/master").WithOperator(corev1.TolerationOpExists).WithEffect(corev1.TaintEffectNoSchedule),
+	}
+
+	podSpec.Volumes = []applycorev1.VolumeApplyConfiguration{
+		*applycorev1.Volume().WithName("keys").WithSecret(applycorev1.SecretVolumeSource().WithSecretName("dr-ssh")),
+	}
+
+	pod := applycorev1.Pod(name, namespace).WithSpec(podSpec)
+	return runPodAndWaitForSuccess(oc, pod)
+}
+
+func runClusterRestoreScript(oc *exutil.CLI, restoreNode *corev1.Node, backupNode *corev1.Node) error {
+	const name = "etcd-restore-pod"
+	const namespace = "openshift-etcd"
+	framework.Logf("running restore script on node: %v", restoreNode.Name)
+
+	backupInternalIp := internalIP(backupNode)
+	restoreInternalIp := internalIP(restoreNode)
+
+	restoreScript := fmt.Sprintf(`
+		#!/bin/bash
+        set -exuo pipefail
+		
+		# ssh key dance
+		%s
+
+        TARGET_NODE_NAME=%s 
+        ssh -i $P_KEY -o StrictHostKeyChecking=no -q core@${TARGET_NODE_NAME} << EOF
+         sudo rm -rf /home/core/backup
+         scp -o StrictHostKeyChecking=no -r core@%s:/home/core/backup .
+         sudo chown -R core /home/core/backup
+         SNAPSHOT=\$(ls -vd /home/core/backup/snapshot*.db | tail -1)
+         mv \$SNAPSHOT /home/core/backup/snapshot.db
+
+         # TODO(thomas): why not use cluster-restore.sh directly? 
+         # the script was designed to create a single recovery etcd instance. Replacing a single node would
+         # create two etcd clusters, which causes some major disruptions after the operation finishes.
+         # Thus below we are manually doing a simple restore with snapshot restore directly. 
+         # This will delete the etcd data dir and the etcd static pod.
+         sudo -s -- <<EOF
+            ls -la /etc/kubernetes/manifests/
+            mv /etc/kubernetes/manifests/etcd-pod.yaml /tmp/etcd.pod.yaml
+            # TODO(thomas): wait for containers to stop?
+
+            source /etc/kubernetes/static-pod-resources/etcd-certs/configmaps/etcd-scripts/etcd.env
+            source /etc/kubernetes/static-pod-resources/etcd-certs/configmaps/etcd-scripts/etcd-common-tools
+            dl_etcdctl
+            rm -rf /var/lib/etcd
+            mkdir -p /var/lib/etcd 
+            etcdctl snapshot restore /home/core/backup/snapshot.db --data-dir=/var/lib/etcd
```

Comment:
I'm probably missing some context but wouldn't doing an `etcdctl snapshot restore` create the new member with a new member id and cluster id?
https://etcd.io/docs/v3.5/op-guide/recovery/#restoring-a-cluster
> Restoring overwrites some snapshot metadata (specifically, the member ID and cluster ID); the member loses its former identity. This metadata overwrite prevents the new member from inadvertently joining an existing cluster. Therefore in order to start a cluster from a snapshot, the restore must start a new logical cluster.

So after this restore step we have the older 2 node cluster and a new 1 node cluster right?
So we have the same issue as the recovery script?

If that is the case, then I'm wondering if we can't just stop the static pods on the non-recovery nodes first. And then restore using the script for the recovery host to get a single node etcd cluster. And then delete and recreate the machines.

Reply:
It depends on what we want to test here. The existing test, this PR disables, does exactly what you mention: 
* Take backup of CP Node 0
* Stop all static pods on non-recovery nodes (Node 0+2)
* Restore backup on CP Node 1, that brings up a single node etcd from snapshot
* Restore API server pods on Node 0+2

Then wait for everything to settle.

What I wanted to start with is a little simpler, akin to reverse vertical scaling:
* Take backup of CP Node 0
* Stop etcd on Node 1, restore from snapshot, restart

It's more similar to the "Replacing an unhealthy etcd member" docs.

https://docs.openshift.com/container-platform/4.12/backup_and_restore/control_plane_backup_and_restore/replacing-unhealthy-etcd-member.html

> So after this restore step we have the older 2 node cluster and a new 1 node cluster right?
> So we have the same issue as the recovery script?

Spot on. I'm simply missing to remove the etcd member before restoring. Gotta figure out how to do that from the terminal (guess with crictl exec or so..)

Reply:
thank god there was a helper for that already. I adjusted the name of the test to be more clear:

> 

also need to roll a new revision out, which is this step in the docs:

> oc patch etcd cluster -p='{"spec": {"forceRedeploymentReason": "recovery-'"$( date --rfc-3339=ns )"'"}}' --type=merge

otherwise it's stuck because the installer installed a newer revision before the restore script put back an older one (nasty)

Reply:
> What I wanted to start with is a little simpler, akin to reverse vertical scaling:
> - Take backup of CP Node 0
> - Stop etcd on Node 1, restore from snapshot, restart

> It's more similar to the "Replacing an unhealthy etcd member" docs.

IIUC it seems like we're scaling down the old cluster and restoring and scaling up the new one serially. Although if that's the case then what happens when we restore and restart the 2nd member of the new cluster?
Does it automatically get added to the new cluster? I'm guessing it shares the same configuration as the other member since etcd snapshot restore would also write the WAL that has the configuration.

> It depends on what we want to test here.

Ultimately we would want to test recovery from a quorum loss scenario like the old test which would mean an unavailable control-plane. Currently we're relying on a healthy control-plane to schedule the `etcd-restore-pod` but I think eventually we would need to account for that not being the case. Which from my understanding was why we have the ssh keys setup so we could do that. Or run it as a static pod.

Reply:
> IIUC it seems like we're scaling down the old cluster and restoring and scaling up the new one serially.

Nope. It's still the same cluster. The "crashing" member is removed and replaced with the same machine, but starting off from a snapshot of another machine.

> Does it automatically get added to the new cluster?

CEO does that automatically, yes. A node with an etcd pod comes up, it's not in the member list and it will be added to the cluster again. Works like a charm, if you disable the quorum guard obviously :sweat_smile: 

> Ultimately we would want to test recovery from a quorum loss scenario like the old test which would mean an unavailable control-plane.

Agreed, let's add this in a follow-up PR. I just wanted to have an end2end that restores one of our easier scenarios to see if the ssh stuff works (it does).

Reply:
Sounds good. We can focus on the quorum loss scenario in the next follow up to see how we want to coordinate that from within a static restore pod.